File: 2025/02/prog_42583_다리를지나는트럭.py
'''
<조건>
- 트럭은 1초에 다리 길이 1씩 전진함 
- 트럭은 올라갈 수 있으면 1초에 한 대씩 다리에 올라갈 수 있음 
- 다리 길이는 bridge_length임 

case 1
- 다리 길이 = 다리에 올라갈 수 있는 최대 대수 = 한 대가 지나야 하는 칸 수 = 2 
1. 다리 위: -, front: 7, pop, 다리 위: (7, 1)
2. 다리 위: 7, front: 4, 최대 무게 초과하므로 pop하지 않음, 다리 위: (7, 2)
3. 다리 위: -, front: 4, pop, 다리 위: (4, 1)
4. 다리 위: (4, 1), front: 5, pop, 다리 위: (4, 2), (5, 1)
5. 다리 위: (5, 1), front: 6, 무게 초과, 다리 위: (5, 2)
6. 다리 위: -, front: 6, pop, 다리 위: (6, 1)
7. 다리 위: (6, 1), 다리 위: (6, 2)
8. 트럭 이동 완료 
'''
from collections import deque
def solution(bridge_length, weight, truck_weights):
    def update_truck(total_weight):
        # 다리 위 트럭 시간 갱신: 1초씩 증가 
        for _ in range(len(bridge_q)):
            truck, time = bridge_q.popleft()
            
            # 트럭 제거 
            if time + 1 == bridge_length:
                total_weight -= truck
            else:
                bridge_q.append((truck, time+1))
    
        return total_weight
    
    answer = 1
    total_weight = 0 # 다리 위에 있는 무게 총합 
    wait_q = deque(truck_weights) # 대기 트럭 
    bridge_q = deque() # 다리 위 트럭 
    limit_weight = weight # 변수 이름 변경 
    
    # 대기 큐와 다리 큐 모두에 트럭이 없으면 종료 
    while wait_q:
        front = wait_q.popleft()
        
        # 최대 무게를 초과한다면 다리로 올리지 않음 
        if total_weight + front > limit_weight:
            wait_q.appendleft(front)
        else:
            # 다리 위로 올림  
            bridge_q.append((front, 0))
            total_weight += front
        
        # 다리 위 트럭 시간 갱신(1초씩 증가) 및 트럭 제거
        total_weight = update_truck(total_weight)
        answer += 1
        
    # 다리 큐 남아있다면 다리 위 남은 트럭 처리 
    while bridge_q:
        total_weight = update_truck(total_weight)
        answer += 1
    
    return answer

File: 2025/07/prog_42583_다리를지나는트럭.js
/*
다리를 건넌다 = 다리 위에 있다 

각 트럭은 다리를 건너는 데에 시간이 얼마나 걸리나?? 
=> bridge_length만큼 
=> 다리 길이 = bridge_length
=> 한 번에 bridge_length개만큼만 
=> 다리는 총 bridge_length칸 
=> 트럭 하나는 다리를 건널 때 bridge_length칸을 지나야 함 = bridge_length 시간이 걸림 

다리 = 큐

순차적으로 큐에 트럭 넣기(트럭무게, 다리에 머물러있는 시간 카운트 값)
매 시간마다 다음 트럭이 다리에 올라올 수 있는지 체크 

while 대기트럭 
    // 1시간에는 기존 삭제와 동시에 새 삽입이 가능함 
    if 큐의 front 카운트가 최대개수와 같다면 
        큐에서 삭제
        현재개수와 현재무게 갱신

    if 다리에 새 트럭을 넣었을 때 다리 최대개수, 최대무게를 모두 넘지 않는다면
        큐에 삽입
        대기트럭열에서 삭제
        현재개수와 현재무게 갱신
    
    전체 시간 카운트 + 1
    
    while 큐
        각 트럭 시간 카운트 + 1

// 다리 위 남은 트럭 처리하기 
// 다리 위 맨 뒤의 트럭이 앞으로 걸리는 시간이 다리가 비어지는 데 걸리는 시간임 
전체 시간 += (다리길이 - 큐의 마지막 원소의 시간)
*/
function solution(bridge_length, weight, truck_weights) {
    let bridgeQ = []
    let totalCnt = 0
    let totalWeight = 0
    let time = 0
    
    while (truck_weights.length > 0) {
        // 다리 맨 앞 트럭 제거 
        if (bridgeQ.length > 0) {
            let [frontWeight, frontTime] = bridgeQ[0]
            if (bridgeQ[0][1] === bridge_length) {
                bridgeQ.splice(0, 1)
                totalCnt -= 1
                totalWeight -= frontWeight
            }
        }
        
        let newTruck = truck_weights[0]
        
        // 새 트럭 추가 
        if ((totalCnt + 1) <= bridge_length && (totalWeight + newTruck) <= weight) {
            // [트럭무게, 다리에 머문 시간]
            bridgeQ.push([newTruck, 0])
            truck_weights.splice(0, 1)
            totalCnt += 1
            totalWeight += newTruck
        }
        
        time += 1
        
        bridgeQ.forEach((_, i) => {
            bridgeQ[i][1] += 1
        })
    }
    
    // 대기 트럭에 남은 트럭은 없지만, 다리 위에는 남은 트럭이 있음 
    time += (bridge_length - bridgeQ[bridgeQ.length-1][1])
    
    return time+1; // 다리가 완전히 비어진 시점도 카운트하기 때문에 + 1
}